keps/sig-node/1665-grpc-probe/README.md

```diff
@@ -0,0 +1,156 @@
+# KEP-1665: Add GRPC Probe
+
+<!-- toc -->
+- [Release Signoff Checklist](#release-signoff-checklist)
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  - [Risks and Mitigations](#risks-and-mitigations)
+- [Design Details](#design-details)
+  - [Followup work (or optionally part of this)](#followup-work-or-optionally-part-of-this)
+  - [Test Plan](#test-plan)
+  - [Graduation Criteria](#graduation-criteria)
+  - [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+  - [Version Skew Strategy](#version-skew-strategy)
+- [Implementation History](#implementation-history)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+<!-- /toc -->
+
+
+## Release Signoff Checklist
+
+- [ ] Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)
+- [ ] KEP approvers have approved the KEP status as `implementable`
+- [ ] Design details are appropriately documented
+- [ ] Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [ ] Graduation criteria is in place
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+GRPC probe - one more opportunity for define container status. Have grpc have official health check protocol: 
+https://github.com/grpc/grpc/blob/master/doc/health-checking.md
+
+## Motivation
+
+GRPC - popular protocol for implement services. For implement health check most of developer use not official solution like:
+https://github.com/grpc-ecosystem/grpc-health-probe
+
+### Goals
+
+Provide for k8 possability for have grpc probe natively(from box) 
+
+### Non-Goals
+
+
+## Proposal
+
+Host: pod ip
+
+```shell script
+    readinessProbe:
+      grpc:
+        port: 9090
```

Comment:
https://github.com/grpc/grpc/blob/master/doc/health-checking.md

That describes being able to set the service name.   Should this support that, too?  With a default of "grpc.health.v1.Health" or whatever the grpc default is?  Or maybe the default should be "" meaning "detect best" ?

Reply:
I think this make no sense in probe, because if pod contain multiple service, you cant restart just one, because prober affecting on all pod @thockin.

Reply:
It's just about the name.  The GRPC spec says clients can query health on a different GRPC service name.  I am just suggesting you expose that here.

Reply:
This comment is still pending

Reply:
+1 on this -- we should model the gRPC health check protocol unless there is a good reason to omit it.